### ISSUE TITLE
TSC cannot Infer Shape of JS Object Literal

### ISSUE THREAD

**Issue by ZheyangSong** [Microsoft/TypeScript#20635]
**TypeScript Version:**  2.5.3

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.

interface IQueryTimeRange {
  start: string | number;
  end: string | number;
}

interface IGetPostBodyParams {
  table: string;
  timeRange: IQueryTimeRange;
  select: string[];
  where?: any[];
  others?: object;
}

interface IPOSTRequestPayload {
  table: string;
  start_time: number | string;
  end_time: number | string;
  select_field: string[];
  where: any[];
  [queryKey: string]: any;
}

function getPostBody({
  table,
  timeRange,
  select,
  where = [
    [{
      name: 'name',
      op: 7,
      value: '',
    }],
  ],
  others = {},
}: IGetPostBodyParams): IPOSTRequestPayload {
  return {
    table,
    start_time: timeRange.start,
    end_time: timeRange.end,
    select_field: select,
    where,
    ...others,
  };
}
```

**Expected behavior:**
TSC shouldn't complain error of incompatible types.

**Actual behavior:**
TSC complain on `return` statement by saying that `Type '{}' is not assignable to type 'IPOSTRequestPayload'.
  Property 'table' is missing in type '{}'.`

**Comment by RyanCavanaugh** [comment]
@sandersn seems like the spread is messing this up?

**Comment by sandersn** [comment]
Yeah, it looks like spread's handling of `object` is still wrong. It should just be skipped, just like string, number, etc.

**Comment by andy-ms** [comment]
You could write `others?: Partial<IPOSTRequestPayload>`, since that seems to be what you intend.

**Comment by ZheyangSong** [comment]
@andy-ms Interesting... didn't try that, syntactically should work but changes my example code's intention. The *others* represents any properties not specified (of course, specified properties are still valid in the `others` object...) by `IPOSTRequestPayload`. In stead of `object`, I guess indexed object could be better... But the issue here is that when the returned object literal obviously has all required properties from `IPOSTRequestPayload`, but TSC still complains...:)

**Comment by andy-ms** [comment]
@ZheyangSong It should be fixed in `typescript@next` now.
